### ISSUE TITLE
Double Pendulum: Purpose of this document

### ISSUE THREAD

**Issue by cd155** [JacquesCarette/Drasil#2631]
In the introduction for the double pendulum, add a section about the purpose of this document.

**Comment by cd155** [comment]
Daily Update:
1. what did today?
I added IPurpose in mkSRS, but it occurs some errors when I try to run the example. The error when it called ChuckDB. The error says "algorithm not found in TermMap". So I went to TermMap to add the algorithm parameter. After fixing the first error, another error occurred says "Reference: parnasClements1986 not found in ReferenceMap".
2. what will do tomorrow?
I will look at the reference error and continue to add the purpose of the document in the double pendulum example.

**Comment by JacquesCarette** [comment]
Could you please work with someone else on the team who has done this before who can help with such errors? I think both @tingyuw and @balacij could do that. The "How To" guide should also be updated with information to help newcomers with such issues.

**Comment by balacij** [comment]
Ah, yes, @cd155 and I spoke about this yesterday over Teams.

Having some information on the "How To" guide is probably good. I'll try to create a starting point.

**Problem Analysis**
When errors are thrown from Drasil programs, we will be given specific error messages in the logs. If the error message isn't familiar to you, it's best to generally search the entire code repository for that message to see what kind of error(s) could have occurred. Using [`grep`](https://linux.die.net/man/1/grep), we can search through the repo fairly easily!

Using `grep "not found in" -r * --include *.hs` (meaning that I want to traverse all files, looking only through the `.hs` files for lines containing "X") shows us that the errors starts in:
https://github.com/JacquesCarette/Drasil/blob/9d2f73e15b69453c8ec3c8203ff1fb4dd837662b/code/drasil-database/Database/Drasil/ChunkDB.hs#L92-L99

`uMapLookup` is the function that _actually_ throws the error, but it's thrown only when a lookup fails for some map (for example, the `symbResolve` function resolves `QuantityDict`s from the `SymbolMap` of a `ChunkDB` using `uMapLookup`, so it can also cause errors).

Looking around the surrounding code (which I won't add to the snippet above for the sake of conserving space), we can see that when "X not found in Y-Map" errors occur, it's usually due to something not being in _some_ `ChunkDB` (the `X` represents the UID of the missing item, and the `Y` represents the Map from which it is expected but does not exist in). 

*Now, what is the `ChunkDB`?*

Using a similar `grep` to the above, we find:

https://github.com/JacquesCarette/Drasil/blob/9d2f73e15b69453c8ec3c8203ff1fb4dd837662b/code/drasil-database/Database/Drasil/ChunkDB.hs#L149-L166

 The `ChunkDB` is the "Chunk Database", containing all "chunk definitions and concepts" for a specific "system" (example). It contains many kinds of `Map`s from UIDs to various chunks.

*So, what does this ultimately mean?*

This means that some example (likely the named one throwing the error in the first place) is missing some chunk information in it's `ChunkDB`!

**Solution**

To resolve this, we generally want to:
1. *Find where the thing we're missing was created* -- Again, I like to do this using `grep` using `grep "X" -r * --include *.hs`. Taking the example of `parnasClements1986`, we get:
```
jasonbalaci@pop-os:~/Programming/Drasil/code$ grep "parnasClements1986" -r * --include *.hs
drasil-data/Data/Drasil/Citations.hs:  parnasClements1986, smithLai2005, lineSource, pointSource :: Citation
drasil-data/Data/Drasil/Citations.hs:parnasClements1986 = cArticle [dParnas, pcClements] 
drasil-data/Data/Drasil/Citations.hs:  "parnasClements1986"
drasil-docLang/Drasil/Sections/Introduction.hs:import Data.Drasil.Citations (parnasClements1986)
drasil-docLang/Drasil/Sections/Introduction.hs:  S "Parnas and Clements point out", refS parnasClements1986 `sC`
drasil-example/Drasil/GlassBR/References.hs:import Data.Drasil.Citations (campidelli, koothoor2013, smithLai2005, parnasClements1986)
drasil-example/Drasil/GlassBR/References.hs:  astm2012, beasonEtAl1998, parnasClements1986]
drasil-example/Drasil/SSP/References.hs:import Data.Drasil.Citations (jnlCGJ, koothoor2013, parnasClements1986, smithLai2005)
drasil-example/Drasil/SSP/References.hs:citations = [chen2005, parnasClements1986, koothoor2013,
drasil-example/Drasil/NoPCM/References.hs:  parnasClements1986, smithLai2005)
drasil-example/Drasil/NoPCM/References.hs:citations = [incroperaEtAl2007, koothoor2013, lightstone2012, parnasClements1986, smithLai2005]
drasil-example/Drasil/GamePhysics/References.hs:import Data.Drasil.Citations (cartesianWiki, koothoor2013, parnasClements1986,
drasil-example/Drasil/GamePhysics/References.hs:citations = [parnas1978, sciComp2013, chaslesWiki, parnasClements1986,
drasil-example/Drasil/SWHS/References.hs:  parnasClements1986, smithLai2005, citeRefs) where
drasil-example/Drasil/SWHS/References.hs:import Data.Drasil.Citations (koothoor2013, parnasClements1986, smithLai2005)
drasil-example/Drasil/SWHS/References.hs:citations = [bueche1986, incroperaEtAl2007, koothoor2013, lightstone2012, parnasClements1986, 
```

Here, we can see that `parnasClements1986` was created in `drasil-data/Data/Drasil/Citations.hs` (we also see some examples of imports of it, so that will be helpful later too).

If no suitable results appeared, then it's likely that you will need to create it, or that a manually written UID reference was incorrect.
2. *Find where we were supposed to place it in the `ChunkDB` of that system/example* -- For example, with DblPendulum, we instantiate our `ChunkDB` here: https://github.com/JacquesCarette/Drasil/blob/9d2f73e15b69453c8ec3c8203ff1fb4dd837662b/code/drasil-example/Drasil/DblPendulum/Body.hs#L138-L145
3. *Figure out which list we should placing it in* -- You will need to analyze the error message for which Map it is related to, and then you will need to read the smart constructor for `cdb`:
https://github.com/JacquesCarette/Drasil/blob/9d2f73e15b69453c8ec3c8203ff1fb4dd837662b/code/drasil-database/Database/Drasil/ChunkDB.hs#L182-L188

The type signature + surrounding comments are very helpful for this.
4. *Add it to that chunk map* -- this heavily depends on the example, but we usually can just prepend it to the list of references. In this case, there is an `allRefs` list inside of `Body.hs` too (which could probably be moved to a new `References.hs` file, like the other examples) for which we can add it.

**Comment by balacij** [comment]
@Ant13731 I think we also came across this issue before too. Do you think the above instructions are ok? If so, do you think we can also add it to the https://github.com/JacquesCarette/Drasil/wiki/Creating-Your-Project-in-Drasil (or perhaps to a FAQ page?)?

**Comment by Ant13731** [comment]
Yea I think that's a great idea! I think your explanation of the chunk database is good as well, maybe we could make a separate page for that? And I only have a small bit to add for this particular case:

The necessary `References` that need to be implemented into the chunk database are usually gathered at the bottom of each example file. It should be separated from the rest of the functions in the file by a `-- References --` header. As of now, the user has to manually add each reference used to the list of references, then import and gather all the separate lists into one large list (`allRefs`) which is then put into the `ChunkDB`. So `allRefs` gathers all the references for each section (or file name):
```Haskell
allRefs :: [Reference]
allRefs = nub (assumpRefs ++ bodyRefs ++ figRefs ++ goalRefs ++ dataDefRefs ++ genDefRefs
  ++ iModRefs ++ tModRefs ++ citeRefs ++ reqRefs ++ secRefs)
```
where each section reference function is defined at the bottom of their respective files. So in `Assumptions.hs`, you'll find `assumpRefs`:
```Haskell
-- References --
assumpRefs :: [Reference]
assumpRefs = map ref assumptions
```

For `DblPendulum` specifically, since it doesn't have a separate `References.hs` file, the `citations` function is defined in `Body.hs`. To fix the error, you'll have to import and add `parnasClements1986` to the existing list of citations:
```Haskell
citations :: BibRef
citations = [accelerationWiki, velocityWiki, hibbeler2004, cartesianWiki]
```

**Comment by Ant13731** [comment]
@balacij do you think having a Wiki page called `Debugging in Drasil` would be better? Then we could link it elsewhere (perhaps in the contributor's guide and creating your project pages)? We could always create an FAQ page afterward and link it there as well.

**Comment by cd155** [comment]
@balacij @Ant13731 Thanks for the start point, the double pendulum example has a slightly different format than other examples, eg physics game. It doesn't have its own reference.hs file. In this ticket, I will create a references.hs file for the double pendulum. With regard to "How To guide", those explanations are really helpful.

**Comment by Ant13731** [comment]
Okay, I think that's a good idea. Just remember to move `citeRefs` to the new `Reference.hs` file and import that function into `Body.hs`

**Comment by cd155** [comment]
@Ant13731 no problem. 

With regard to `allRefs` that @balacij mentioned in Solution point 4, I checked with most examples, `allRefs` seems to belong to body.hs, as most examples have it on body.hs.

**Comment by Ant13731** [comment]
I have created a new page [Debugging in Drasil](https://github.com/JacquesCarette/Drasil/wiki/Debugging-in-Drasil) that records the guide from @balacij. Please let me know if I should add anything.

**Comment by balacij** [comment]
Ah, thanks @Ant13731, that's a great idea! I see some good changes too (Haddock is good!).

Do you mind using the permalinks how I used them in my original post? Permalinks carry more information (the file location + commit hash is particularly helpful, so they know what approximately when the doc was written during) and I assumed that readers would note the files while reading.

**Comment by Ant13731** [comment]
The permalinks don't seem to render automatically like they do in issues. I couldn't find anything in the permalink documentation either. I can still add the links to the file names?

**Comment by balacij** [comment]
Oh, that's weird. I didn't know they had different renderers on the wiki vs the comments. That's good to know!

I can add them, no worries. We can probably also add a bit more information about the issue I think. Thank you!

**Comment by balacij** [comment]
Ah, it looks like you already fixed them before me, thank you @Ant13731 :smile: ! The doc is looking great!

PS: I've also started using [`ripgrep`](https://github.com/BurntSushi/ripgrep#installation) as a replacement to `grep`. It's worth a try if you're interested!

**Comment by smiths** [comment]
Great discussion @balacij, @Ant13731 and @cd155.  I really like that you captured the pertinent parts of the discussion in the debugging in Drasil Wiki.  Gold!

I didn't test everything on my Mac, but I did test the `grep` instructions.  The `grep` instructions as given didn't actually work for the Mac.

```grep "not found in" -r * --include *.hs```

needs to be rewritten with quotes in the `--include`:

```grep "not found in" -r * --include "*.hs"```

I haven't investigated, but I'm using the `grep` that ships with the Mac.  I first tried with the default `grep` that ships with the Mac.  I then tried with the homebrew version, `ggrep`.  I needed the quotes in both cases.

Would it make sense to change the Debugging in Drasil instructions to use the quotes?  I haven't tested on my Linux VM, but I'm optimistic that the quotes will work on Linux.  Alternatively, we could just give a warning in the instructions that the Mac version will be different, since it will need quotes?

**Comment by smiths** [comment]
@balacij, I had a look at `ripgrep`.  Would the equivalent of the above be?

```rg "not found in" -ths```

**Comment by balacij** [comment]
Ah, good catch! Thank you! I think it would be best to just write it with the double quotes. I'll update it.

That would be a slight upgrade actually. Running `rg --type-list` gives the following:
```
...
haskell: *.c2hs, *.cpphs, *.hs, *.hsc, *.lhs
hbs: *.hbs
hs: *.hs, *.lhs
...
```

So, it looks like yours would also catch the lhs files (which is good). I really like ripgrep respecting our .gitignores out of the box.

**Comment by smiths** [comment]
Yes, ripgrep seems like a nice tool.  I think it is also recursive by default.

**Comment by JacquesCarette** [comment]
This issue got closed when I merged in #2665  - but I want to double-check that all the super-useful information above, on creating and debugging, did indeed make it to the wiki?

**Comment by Ant13731** [comment]
Yep, I think all the really important stuff is in the [wiki](https://github.com/JacquesCarette/Drasil/wiki/Debugging-in-Drasil). We could include a little bit more about `ripgrep` (right now we only have a `grep` section).
